api/integrations/github/github.py

```diff
@@ -0,0 +1,129 @@
+import logging
+import typing
+
+import requests
+from django.conf import settings
+
+from features.models import FeatureState, FeatureStateValue
+from integrations.github.client import generate_token
+from integrations.github.constants import GITHUB_API_URL
+from integrations.github.models import GithubConfiguration
+from webhooks.webhooks import WebhookEventType
+
+logger = logging.getLogger(__name__)
+
+
+def post_comment_to_github(
+    installation_id: str, owner: str, repo: str, issue: str, body: str
+) -> typing.Optional[typing.Dict[str, typing.Any]]:
+    try:
+        token = generate_token(
+            installation_id,
+            settings.GITHUB_APP_ID,
+        )
+
+        url = f"{GITHUB_API_URL}repos/{owner}/{repo}/issues/{issue}/comments"
+        headers = {
+            "Accept": "application/vnd.github.v3+json",
+            "Authorization": f"Bearer {token}",
+        }
+
+        payload = {"body": body}
+        response = response = requests.post(
+            url, json=payload, headers=headers, timeout=10
+        )
+
+        return response.json() if response.status_code == 200 else None
+    except requests.RequestException as e:
+        logger.info(f" {e}")
+        return None
+
+
+def generate_body_comment(
+    name: str,
+    event_type: str,
+    feature_states: typing.List[typing.Dict[str, typing.Any]],
+) -> str:
+
+    is_update = event_type == WebhookEventType.FLAG_UPDATED.value
+    is_removed = event_type == WebhookEventType.FEATURE_EXTERNAL_RESOURCE_REMOVED.value
+    delete_text = f"### The feature flag {name} was unlinked from the issue/PR"
+
+    if is_removed:
+        return delete_text
+
+    last_updated_string = (
+        f"Last Updated {datetime.datetime.now().strftime('%dth %b %Y %I:%M%p')}"
+    )
+    updated_text = (
+        f"### The Flagsmith Feature {name} " f"was updated in the environment "
+    )
+
+    result = (
+        updated_text
+        if is_update
+        else f"### This pull request is linked to a Flagsmith Feature ({name}):\n"
+    )
+
+    for v in feature_states:
+        feature_value = v.get("feature_state_value")
+        feature_value_type = v.get("feature_state_value_type")
+
+        feature_value_string = f"\n{feature_value_type}\n```{feature_value if feature_value else 'No value.'}```\n\n"
+
+        result += f"**{v['environment_name']}{' - ' + v['segment_name'] if v.get('segment_name') else ''}**\n"
+        result += f"- [{'x' if v['feature_value'] else ' '}] {'Enabled' if v['feature_value'] else 'Disabled'}"
+        result += feature_value_string
+
+    result += last_updated_string
+    return result
+
+
+def check_not_none(value: any) -> bool:
+    return value is not None
+
+
+def generate_data(
+    github_configuration: GithubConfiguration,
+    feature_id: int,
+    feature_name: str,
+    type: str,
+    feature_states: typing.Union[list[FeatureState], list[FeatureStateValue]] = None,
+    url: str = None,
+) -> dict[str, typing.Any]:
```

Comment:
Can we convert the response this functions returns to a data class?

Reply:
Using a data class now